cli/script/command-executor.ts

```diff
@@ -912,25 +913,57 @@ function getReactNativeProjectAppVersion(command: cli.IReleaseReactCommand, proj
             throw new Error("Unable to find or read \"build.gradle\" in the \"android/app\" folder.");
         }
 
+        const validVersion = (version: string) => semver.valid(version) || missingPatchVersionRegex.test(version);
+        
         return g2js.parseFile(buildGradlePath)
-            .catch((err: Error) => {
-                throw new Error("Unable to parse the \"android/app/build.gradle\" file, it could be malformed.");
+            .catch(() => {
+                throw new Error(`Unable to parse the "android/app/build.gradle" file. Please ensure it is a well-formed Gradle file.`);
             })
             .then((buildGradle: any) => {
-                if (buildGradle.android && buildGradle.android.defaultConfig && buildGradle.android.defaultConfig.versionName) {
-                    if (typeof buildGradle.android.defaultConfig.versionName !== "string") {
-                        throw new Error(`The "android.defaultConfig.versionName" property value in "android/app/build.gradle" is not a valid string. If this is expected, consider using the --targetBinaryVersion option to specify the value manually.`);
-                    }
+                if (!buildGradle.android || !buildGradle.android.defaultConfig || !buildGradle.android.defaultConfig.versionName) {
+                    throw new Error(`The "android/app/build.gradle" file doesn't specify a value for the "android.defaultConfig.versionName" property.`);
+                }
 
-                    var appVersion: string = buildGradle.android.defaultConfig.versionName.replace(/"/g, "").trim();
-                    if (semver.valid(appVersion) || missingPatchVersionRegex.test(appVersion)) {
-                        return appVersion;
-                    } else {
-                        throw new Error("The \"android.defaultConfig.versionName\" property in \"android/app/build.gradle\" needs to have at least a major and minor version, for example \"2.0\" or \"1.0.3\".");
-                    }
-                } else {
-                    throw new Error("The \"android/app/build.gradle\" file does not include a value for android.defaultConfig.versionName.");
+                if (typeof buildGradle.android.defaultConfig.versionName !== "string") {
+                    throw new Error(`The "android.defaultConfig.versionName" property value in "android/app/build.gradle" is not a valid string. If this is expected, consider using the --targetBinaryVersion option to specify the value manually.`);
+                }
+
+                let appVersion: string = buildGradle.android.defaultConfig.versionName.replace(/"/g, "").trim();
+
+                if (validVersion(appVersion)) {
+                    return appVersion;
                 }
+
+                // The version property isn't a valid semver string
+                // so we assume it is a reference to a property variable.
+
+                const propertyName = appVersion.replace("project.", "");
```

Comment:
Also, the properties that are defined via `grade.properties` are exposed to the build file as properties on the `project` object, so I don't think we need to make this solution any more generic, since there isn't another global object that property references could be made off of (e.g. `build.FOO`).